### PR TITLE
Add ExecutionThreadApartmentState property in runsettings

### PR DESCRIPTION
**Pull request by smadala**
- Provide a way to specify thread apartment state for thread which calls adapters ExecuteTests(), Cancel APIs.
More details: https://github.com/Microsoft/vstest-docs/pull/64

### CODE REVIEW

File: test/TestAssets/SimpleTestProject3/UnitTest1.cs [hunk 1]
```diff
@@ -31,5 +34,27 @@ public void ExitWithStackoverFlow()
...
+        {
+            Clipboard.SetText("Clipboard");
+            Thread.Sleep(1000 * 5);
+        }
+
+        [TestMethod]
+        public void UITestWithSleep2()
+        {
+            Clipboard.SetText("Clipboard");
+            Thread.Sleep(1000 * 10);
```
Comment (codito) [inline test/TestAssets/SimpleTestProject3/UnitTest1.cs]: Ouch.

**Review by codito — DISMISSED** [review]

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs [hunk 2]
```diff
@@ -94,7 +110,28 @@ internal abstract class BaseRunTests
...
+            this(runSettings,
+                testExecutionContext,
+                testCaseEventsHandler,
+                testRunEventsHandler,
+                testPlatformEventSource,
+                testEventsPublisher, new PlatformThread())
+        {
+        }
+
+        internal BaseRunTests(string runSettings,
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: Removed the internal .ctor.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs [hunk 3]
```diff
@@ -344,7 +390,19 @@ private bool RunTestInternalWithExecutors(IEnumerable<Tuple<Uri, string>> execut
                         this.testPlatformEventSource.AdapterExecutionStart(executorUriExtensionTuple.Item1.AbsoluteUri);
 
                         // Run the tests.
-                        this.InvokeExecutor(executor, executorUriExtensionTuple, this.runContext, this.frameworkHandle);
+                        if (RequiredSTAThread())
+                        {
+                            // Call adapter ExecuteTests in thread with STA.
+                            // Tests need to be run in STA thread when accessin COM objects.
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs]: Removed the comment.

File: src/Microsoft.TestPlatform.ObjectModel/Utilities/XmlRunSettingsUtilities.cs [hunk 4]
```diff
@@ -264,6 +264,73 @@ public static DataCollectionRunSettings GetDataCollectionRunSettings(string runS
...
+
+            return null;
+        }
+
+        /// <summary>
+        /// Get ExecutionThreadApartmentStateKey from runsettings.
+        /// </summary>
+        /// <param name="settingsXml"> The runsettings. </param>
+        /// <returns> Thread Apartment State. </returns>
+        public static PlatformApartmentState GetExecutionThreadApartmentState(string settingsXml)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.ObjectModel/Utilities/XmlRunSettingsUtilities.cs]: Added ExecutionThreadApartmentState prop to RunConfiguration. Removed this method. Not combining the data collector settings reading because not using then always together in other places.

File: src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+{
+    using System;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces;
+
+    public class PlatformThread : IThread
+    {
+        /// <inheritdoc/>
+        public void Run(Action action, PlatformApartmentState apartmentState, bool waitForCompletion)
+        {
+            throw new NotImplementedException();
```
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs]: Fixed this, Now displaying warning and fallbacking to MTA.

File: test/Microsoft.TestPlatform.AcceptanceTests/ExecutionThreadApartmentStateTests.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            var assemblyPaths =
+                this.BuildMultipleAssemblyPath("SimpleTestProject3.dll").Trim('\"');
+            var arguments = PrepareArguments(assemblyPaths, this.GetTestAdapterPath(), string.Empty, this.FrameworkArgValue);
+            arguments = string.Concat(arguments, " /testcasefilter:UITestMethod");
+            this.InvokeVsTest(arguments);
+            this.ValidateSummaryStatus(0, 1, 0);
+        }
+
+        [Ignore]
+        // Issue with cancel, After cancellation same test counting twice. Below is example command and output.
```
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/ExecutionThreadApartmentStateTests.cs]: Done.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Execution/BaseRunTestsTests.cs [hunk 7]
```diff
@@ -588,6 +594,66 @@ public void RunTestsShouldReportLogMessagesFromExecutors()
...
+                    else
+                    {
+                        // Callback for Cancel().
+                        cancelEvent.Set();
+                    }
+                });
+            Task.Run(() => this.runTestsInstance.RunTests());
+            // RunTests should be in progress to do Cancel.
+            // Adding sleep ensures RunTests task executes first.
+            Thread.Sleep(100);
```
Comment (smadala) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Execution/BaseRunTestsTests.cs]: Removed it.

File: test/TestAssets/SimpleTestProject3/UnitTest1.cs (hunk 1 above)
Comment (smadala) [inline test/TestAssets/SimpleTestProject3/UnitTest1.cs]: Removed it.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces;
+
+    public class PlatformThread : IThread
+    {
+        /// <inheritdoc/>
+        public void Run(Action action, PlatformApartmentState apartmentState, bool waitForCompletion)
+        {
+            if (apartmentState == PlatformApartmentState.STA)
+            {
+                throw new NotSupportedThreadApartmentStateException("Currently STA Thread apartment state not supported for .NetCore");
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs]: Should this go in resx?

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/NotSupportedThreadApartmentStateException.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.PlatformAbstractions
+{
+    using System;
+
+    /// <summary>
+    /// Not supported thread aprtment state exception.
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/NotSupportedThreadApartmentStateException.cs]: Nit: apartment
Prefer naming : ThreadApartmentStateNotSupportedException.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            }
+
+            if (action == null)
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs]: When can this happen? Should we log a warning here?

File: test/Microsoft.TestPlatform.AcceptanceTests/ExecutionTests.cs [hunk 11]
```diff
@@ -144,7 +144,7 @@ public void StackOverflowExceptionShouldBeLoggedToConsoleAndDiagLogFile(string r
             var assemblyPaths =
                 this.BuildMultipleAssemblyPath("SimpleTestProject3.dll").Trim('\"');
             var arguments = PrepareArguments(assemblyPaths, this.GetTestAdapterPath(), string.Empty, this.FrameworkArgValue);
-            arguments = string.Concat(arguments, " /tests:ExitWithStackoverFlow");
+            arguments = string.Concat(arguments, " /testcasefilter:ExitWithStackoverFlow");
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.AcceptanceTests/ExecutionTests.cs]: Why is this change required?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Resources/Resources.resx [hunk 12]
```diff
@@ -183,6 +183,9 @@
   <data name="FailedToLaunchTestHost" xml:space="preserve">
     <value>Failed to launch testhost with error: {0}</value>
   </data>
+  <data name="UnableToExecuteInSTAThread" xml:space="preserve">
+    <value>Unable to execute in STA thread, Using MTA thread. Reason: {0}</value>
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Resources/Resources.resx]: Please get a signoff from @pvlakshm

**Review by codito — COMMENTED** [review]
@Faizan2304 can you please review this?

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                {
+                    action();
+                }
+                catch (Exception e)
+                {
+                    exThrown = e;
+                }
+            });
+
+            ApartmentState state = Enum.TryParse(apartmentState.ToString(), out state) ? state : ApartmentState.MTA;
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs]: Should we notify user if we are not running test in STA? Because when we are here, we are under the impression that we are going to run test in STA.

**Review by Faizan2304 — APPROVED** [review]

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IThread.cs [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    public interface IThread
+    {
+        /// <summary>
+        /// Runs the action in a thread with given apartment state.
+        /// </summary>
+        /// <param name="action">The Action to be called. No-Op if action is null.</param>
+        /// <param name="apartmentState">The apartment state.</param>
+        /// <param name="waitForCompletion"> True for running in Sync, False for running in Async</param>
+        /// <exception cref="ThreadApartmentStateNotSupportedException"> When thread apartment state not supported for platform. </exception>>
+        void Run(Action action, PlatformApartmentState apartmentState, bool waitForCompletion);
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IThread.cs]: Should we have a parallel RunAsync API, following the TAP pattern?

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        /// <inheritdoc/>
+        public void Run(Action action, PlatformApartmentState apartmentState, bool waitForCompletion)
+        {
+            if (action == null)
+            {
+                return;
+            }
+
+            Exception exThrown = null;
+            var thread = new System.Threading.Thread(() =>
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs]: We are always creating a new thread, should we check if user has changed apartment state, then only we want to create new thread?

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+                throw new ThreadApartmentStateNotSupportedException();
+            }
+
+            if (action == null)
+            {
+                return;
+            }
+
+            Exception exThrown = null;
+            var thread = new Thread(() =>
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs]: If netcore only supports MTA, then no need to create new thread, we can simply invoke action?

File: src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces;
+
+    public class PlatformThread : IThread
+    {
+        /// <inheritdoc/>
+        public void Run(Action action, PlatformApartmentState apartmentState, bool waitForCompletion)
+        {
+            if (apartmentState == PlatformApartmentState.STA)
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs]: Should we throw exception here? Users generally tend to use same run settings for a session(IDE), although STA won't make sense here, we could simply ignore the apartment state, Log it, & continue running?

**Review by mayankbansal018 — CHANGES_REQUESTED** [review]

File: src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs (hunk 17 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs]: You are right, we are logging the warning and running in MTA in  caller (`BaseRunTests.cs`). As there is no way to log in platform abstraction, we are doing that by throwing exception.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs (hunk 16 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs]: If value is MTA, we don't even call this API. the check has been done in `BaseRunTests.cs`.

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs (hunk 15 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs]: Doing the check in `BaseRunTests.cs`.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs (hunk 8 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs]: Not required anymore, changed the warning message.

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/NotSupportedThreadApartmentStateException.cs (hunk 9 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/NotSupportedThreadApartmentStateException.cs]: Done.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs (hunk 10 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/System/PlatformThread.cs]: Currently there is no scenario now, Updated the API doc to No-Op when action is null.

File: test/Microsoft.TestPlatform.AcceptanceTests/ExecutionTests.cs (hunk 11 above)
Comment (smadala) [inline test/Microsoft.TestPlatform.AcceptanceTests/ExecutionTests.cs]: To avoid launching two testhosts.

File: src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs [hunk 18]
```diff
@@ -301,6 +303,15 @@ public string TestAdaptersPaths
...
         /// <summary>
+        /// Gets or sets the execution thread apartment state.
+        /// </summary>
+        public PlatformApartmentState ExecutionThreadApartmentState
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
```
Comment (smadala) [inline src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs]: Good catch. Done.

File: src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs [hunk 19]
```diff
@@ -724,6 +735,24 @@ public static RunConfiguration FromXml(XmlReader reader)
                             XmlRunSettingsUtilities.ThrowOnHasAttributes(reader);
                             runConfiguration.BinariesRoot = reader.ReadElementContentAsString();
                             break;
+                        case "ExecutionThreadApartmentState":
```
Comment (smadala) [inline src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs]: Done.

File: src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+            {
+                throw new NotSupportedThreadApartmentStateException("Currently STA Thread apartment state not supported for UAP10.0");
```
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/System/PlatformThread.cs]: Not required anymore, changed the warning.

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs (hunk 13 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs]: Notifying on invalid value in RunConfiguration.cs.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Resources/Resources.resx (hunk 12 above)
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Resources/Resources.resx]: Changed the resource string. Got signoff from @pvlakshm.

**Comment by smadala** [comment]
@dotnet-bot test Windows_NT / Release Build please.

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs (hunk 15 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformThread.cs]: nice

**Review by mayankbansal018 — APPROVED** [review]
